flink-table-store-core/src/main/java/org/apache/flink/table/store/file/casting/CastExecutors.java

```diff
@@ -176,7 +176,7 @@ public class CastExecutors {
                 }
                 return null;
             case BINARY:
-                if (outputType.getTypeRoot() == BINARY) {
+                if (outputType.getTypeRoot() == BINARY || outputType.getTypeRoot() == VARBINARY) {
```

Comment:
I think we should not use this code. Here should something like Char and Varchar.

Reply:
Done and add test case